feat(cli,metro-config): Add supervisor transform worker (wrapping user-configured `transformerPath`)

kitten opened the pull request:
# Why

While #38901 provides new entrypoints to allow a stable entrypoint for third-party libraries to customize the Metro transformer and Babel transformer (while we don't have a plugin API yet), this PR aims to improve what happens when a third-party library hasn't been added.

It aims to assign blame directly and clearly to the custom transformer, if an error occurs while loading it. It also redirects modules as needed, to enforce a single transformer version (mismatching `metro`/`metro-transform-worker`/`@expo/metro-config` versions loaded by third-party transformers will always lead to errors)

This specifically targets `transformerPath` and `transformer.babelTransformerPath` customisations that are out-of-date.

The reasoning here is that if we control the "entrypoint" transformer then we can always customise how custom transformers behave.

# How

- Overwrite `transformerPath` with a "supervising" transformer,
  - when the user has changed the `transformerPath`
  - when the user has changed the `transformer.babelTransformerPath`
- Add a `supervising-tranformer-worker` that optionally loads the custom transformer from `config.transformer.expo_customTransformerPath`
- If the transformer fails to require, we throw a custom error that's clearer than Metro's default
- If the transformer requires a module that matches the "known broken" imports, we redirect them to fixed versions

Example of what the initialization error looks like:
<img width="1453" height="78" alt="Screenshot 2025-08-19 at 09 57 55" src="https://github.com/user-attachments/assets/c5eec101-8272-475e-a808-5406e7e02e52" />

# Test Plan

Typical cases this addresses are:
- add a `throw` to the top-level of a custom transformer -> the new error message will be shown
- use a transformer that imports `metro-transform-worker` / `metro` / `@expo/metro-config` -> a redirect will occur
- use a Babel transformer that imports `metro-babel-transformer` -> a redirect will occur
- misaligned dependencies that'd previously cause an error may not cause errors any longer

Manual testing:
- [x] Tested against `https://github.com/nativewind/nativewind/commit/92e43ff`
- [x] Tested against `https://github.com/nativewind/nativewind/commit/92e43ff` with `bun run --bun` (v1.2.20)
- [x] Tested against a `react-native-svg-transformer` project with misaligned `@expo/metro-config` version in root
- [x] Tested against working (no dependency conflicts) project with custom Babel transform requiring `@expo/metro-config/babel-transformer`
- [x] ~~Tested against Nativewind v4 with mismatching `metro-transform-worker` version in root `node_modules`~~

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

kitten committed:
metro-config: Add `unstable_transformerPath` error

kitten committed:
Add superivising-transform-worker

kitten committed:
Rebuild

kitten committed:
Add supervising-transform-worker wrapper

kitten committed:
Apply module remapping in supervising-transform-worker

The supervising transform worker wraps around the user's custom
transformer. When this custom transformer requests modules that match
a limited set of SDK packages that are fixed to the `expo` package (i.e.
the SDK version) we redirect the request to fixed paths.

This ensures that a transform worker doesn't break even if it imports
from "illegal" paths.

This applies to:
- `@expo/metro-config` (self)
- `@expo/metro`
- `metro-transform-worker`
- `metro`

These packages should never mismatch with the SDK's version of them.

kitten committed:
metro-config: Rebuild

kitten committed:
Apply supervising transformerPath

kitten committed:
Add CHANGELOG entries

kitten committed:
Allow supervisor without expo_customTransformerPath and add metro-babel-transformer remap

kitten committed:
metro-config: Rebuild

kitten committed:
Add supervising-transform-worker when Babel tranformer was customized too

kitten committed:
Support mocked config.transformer object

kitten committed:
Add tests for withMetroSupervisingTransformWorker

krystofwoldrich commented on packages/@expo/cli/src/start/server/metro/withMetroSupervisingTransformWorker.ts:
```diff
@@ -0,0 +1,50 @@
+  const originalBabelTransformerPath = config.transformer?.babelTransformerPath;
+
+  const hasOriginalTransformPath = originalTransformerPath === unstable_transformerPath;
+  const hasOriginalBabelTransformerPath =
+    !originalBabelTransformerPath ||
+    defaultBabelTransformerPaths.includes(originalBabelTransformerPath);
```
Nit: Falsy value won't be included in the array of defaults, we could leave the check out. To me it reads easier and don't see any large negatives (- the early bail).

````suggestion
  const hasOriginalBabelTransformerPath = defaultBabelTransformerPaths.includes(
    originalBabelTransformerPath
  );
````

kitten committed:
Refactor and add remapping tests

kitten committed:
metro-config: Rebuild

kitten committed:
Separate out debug messages

kitten committed:
Add full list of metro modules (doesn't really hurt)

krystofwoldrich commented on packages/@expo/cli/src/start/server/metro/withMetroSupervisingTransformWorker.ts:
```diff
@@ -0,0 +1,49 @@
+    return config;
+  }
+
+  if (!hasOriginalBabelTransformerPath && hasOriginalTransformPath) {
```
I might have missed something, but should this one check only `!hasOriginalTransformPath`

kitten replied:
oops

kitten commented on packages/@expo/metro-config/src/transform-worker/supervising-transform-worker.ts:
```diff
@@ -0,0 +1,138 @@
+  // Use defaults for everything that's not custom.
+  ...worker,
+
+  // We ensure that core utilities are never replaced
+  applyImportSupport: worker.applyImportSupport,
+  getCacheKey: worker.getCacheKey,
+  collectDependenciesForShaking: worker.collectDependenciesForShaking,
+  minifyCode: worker.minifyCode,
```
They are, yea. I've added them for clarity and to keep track of what's exported. My original plan was to add a `satisfies` check here and in `transform-worker` too but couldn't get that to work.

I'm listing them so we have a clear list of what's applied from our worker

kitten committed:
Only map exact packages to known good node_modules search path

Prevents resolving wrong/excessive amount of packages

kitten committed:
Rebuild

kitten committed:
Change error to relative path

kitten committed:
Remove `expo_customTransformerPath` from cache key

kitten committed:
Merge branch 'main' into @kitten/feat/transformer-init-errors

EvanBacon reviewed (approved):
wild stuff. Let's add some more comments about how it works and maybe some ability to disable the override so we can debug things a bit easier.

kitten committed:
Add comments and refactor

kitten committed:
metro-config: Rebuild

kitten committed:
Noop explicitly when the default transformer was passed

kitten committed:
Add comment to CLI

kitten committed:
Allow the supervisor to be force-disabled

kitten commented:
Can now be disabled by setting `config.transformer.expo_transformerPath: false`. Not a clear way to disable it since it's just for debugging.

Refactored the module remapping a little and added a ton of comments to explain it, since it's a weird thing to do.

If we see any failures we can't explain and think it's caused by this supervisor, we should just revert this.
